### ISSUE TITLE
Unbounded Transform for integer-to-float formats

### ISSUE THREAD

**Issue by jpap** [mm2/Little-CMS#276]
Thanks for an awesome project!

I am using LCMS to perform a transform from a Display P3 RGB profile to extended range sRGB.  I have noticed that unbounded mode only kicks in when **both** the input and output formats are float.  It would be very nice if unbounded mode works when the input format is integer (and output format is float).  This could be opt-in.

My source content is 8bit/component RGBA.  Currently I have to allocate a source image buffer twice as big and convert the source into a Half-Float image, before executing the LCMS transform to Half-Float output, sRGB extended range.  Having LCMS support integer inputs while maintaining extended range greatly reduces memory usage in this case.

**Comment by jpap** [comment]
The PR given works well for me for 8bit/component RGBA input to (half-)float output.  Appreciate your feedback and happy to make changes so we can land the change.

**Comment by mm2** [comment]
You can do this by using a formatter plug-in, which does not need any modification to core library and preserves all gamut check and other features. 

See below an example for RGB 8 bits. You may want to modify the unroll function for your needs. See cmspack.c for samples of unroll functions. All is documented in the plug-ins manual.

```
#define TYPE_RGB8_AS_FLOAT  (COLORSPACE_SH(PT_RGB)|CHANNELS_SH(3)|BYTES_SH(1)|FLOAT_SH(1))

static
cmsUInt8Number* my_UnrollRGB8AsFloat(CMSREGISTER _cmsTRANSFORM* info,
                                     CMSREGISTER cmsFloat32Number wIn[],
                                     CMSREGISTER cmsUInt8Number* accum,
                                     CMSREGISTER cmsUInt32Number Stride)                                  
{
    wIn[0] = (cmsFloat32Number) accum[0] / 255.0f;
    wIn[1] = (cmsFloat32Number) accum[1] / 255.0f;
    wIn[2] = (cmsFloat32Number) accum[2] / 255.0f;

    return accum + 3;
}

cmsFormatter my_FormatterFactory(cmsUInt32Number Type, 
                                 cmsFormatterDirection Dir,                                                   
                                 cmsUInt32Number dwFlags)
{
    cmsFormatter Result = { NULL };

    if ((Type == TYPE_RGB8_AS_FLOAT) &&
        (dwFlags & CMS_PACK_FLAGS_FLOAT) &&
        (Dir == cmsFormatterInput)) {
        Result.Fmt16 = my_UnrollRGB8AsFloat;
    }
    return Result;
}

static
cmsPluginFormatters FormattersPluginSample = { { 
                                cmsPluginMagicNumber,
                                2060,
                                cmsPluginFormattersSig,
                                NULL},
                                my_FormatterFactory };

...
cmsPlugin(&FormattersPluginSample);
xform = cmsCreateTransform( hsRGB, TYPE_RGB8_AS_FLOAT, hWhateverRGB, TYPE_RGB_DBL, INTENT_PERCEPTUAL, 0);

...

```

**Comment by jpap** [comment]
Thanks for your clever suggestion for using an input formatter plug-in as a workaround.  

Given that the behavior of lcms is different for int vs float input types, when `cmsFLAGS_NONEGATIVES` is not selected and the output format is float, I've suggested in #277 that the PR is better suited in the core library.

Otherwise the lcms library user has to be aware of the above behavior (which is not obvious!) so as to selectively invoke your suggested workaround when unbounded (extended range) output is desired, for integer input and float output formats.  It is not a good idea to always unroll inputs as floats as a rule otherwise, because that'll likely be a negative impact on performance compared to the int->int path when suitable.

Of course the suggested workaround can be encapsulated in a wrapper layer around lcms that is conditionally enabled, but I would argue that is yet another reason for its inclusion into the core library from the start.

**Comment by mm2** [comment]
After inspection, this has proven to be a real issue. 

Solved by 195c218dd3c8338e12563c9641ef89f84cc6f6b4

The behaviour now is unbounded (float) mode is selected automatically if you use a floating point formats on output. Yout need not any special flag to mark those transforms, just use floating point on output. cmsFLAGS_NONEGATIVES works as well.
